python/rascaline/rascaline/utils/clebsch_gordan/correlate_density.py

```diff
@@ -76,185 +88,331 @@ def correlate_density(
+        skip_redundant: Optional[Union[bool, List[bool]]] = False,
+        output_selection: Optional[Union[bool, List[bool]]] = None,
+        arrays_backend: Optional[str] = None,
+        cg_backend: Optional[str] = None,
+    ):
+        super().__init__()
+        if arrays_backend is None:
+            if torch_jit_is_scripting():
+                self._arrays_backend = "torch"
+            else:
+                if isinstance(Labels, TorchScriptClass):
+                    self._arrays_backend = "torch"
+                else:
+                    self._arrays_backend = "numpy"
+        elif arrays_backend == "numpy":
+            if torch_jit_is_scripting():
+                raise ValueError(
+                    "Module is torch scripted but 'numpy' was given as `arrays_backend`"
+                )
+            self._arrays_backend = "numpy"
+        elif arrays_backend == "torch":
+            self._arrays_backend = "torch"
+        else:
+            raise ValueError(
+                f"Unkown `arrays_backend` {arrays_backend}."
+                "Only 'numpy' and 'torch' are supported."
+            )
+
+        # Choosing the optimal cg combine backend
+        if cg_backend is None:
+            if self._arrays_backend == "torch":
+                self._cg_backend = "python-dense"
+            if self._arrays_backend == "numpy" and HAS_MOPS:
+                self._cg_backend = "mops"
+            else:
+                self._cg_backend = "python-sparse"
+        elif cg_backend == "python-dense":
+            self._cg_backend = "python-dense"
+        elif cg_backend == "python-sparse":
+            self._cg_backend = "python-sparse"
+        elif cg_backend == "mops":
+            if self._arrays_backend == "torch":
+                raise NotImplementedError(
+                    "'numpy' was determined or given as `arrays_backend` "
+                    "and 'mops' was given as `cg_backend`, "
+                    "but mops does not support torch backend yet"
+                )
+        else:
+            raise ValueError(
+                f"Unkown `cg_backend` {cg_backend}."
+                "Only 'python-dense', 'python-sparse' and 'mops' are supported."
+            )
+
+        if max_angular < 0:
+            raise ValueError(
+                f"Given `max_angular={max_angular}` negative. "
+                "Must be greater equal 0."
+            )
+        self._max_angular = max_angular
+
+        if self._cg_backend == "python-dense":
+            sparse = False
+            use_mops = False
+        elif self._cg_backend == "python-sparse":
+            sparse = True
+            use_mops = False
+        elif self._cg_backend == "mops":
+            sparse = True
+            use_mops = True
+
+        # We cannot store this into one member variable because TorchScript
+        # has problems scripting which cannot be resolved with isinstance checks
```

Comment:
Some remains of me trying to deal with type `Union[SparseCgDict, DenseCgDict]` and TorchScript. But this is not up to date. Can be removed.
```suggestion
```